internal/controller/grpc_logger.go

```diff
@@ -226,20 +228,23 @@ func NewGRPClogger(grpcSyncer *BufferedGrpcWriteSyncer) *zap.SugaredLogger {
 
 	// Add a custom hook to handle logs for grpcSyncer
 	logger = logger.WithOptions(zap.Hooks(func(entry zapcore.Entry) error {
+		// Do not use logging inside the hook to avoid deadlock
+		var shouldBuffer bool
+
 		grpcSyncer.mutex.Lock()
-		defer grpcSyncer.mutex.Unlock()
 		if grpcSyncer.conn == nil || grpcSyncer.conn.GetState() != connectivity.Ready {
-			grpcSyncer.bufferLogEntry(&entry)
-			return nil
+			shouldBuffer = true
+		} else {
+			// Flush pending logs
+			grpcSyncer.flush()
+			if err := grpcSyncer.sendLog(&entry, nil); err != nil {
+				shouldBuffer = true
+			}
 		}
+		grpcSyncer.mutex.Unlock()
 
-		// Flush buffer so logs are sent in order to the server
-		grpcSyncer.flush()
-
-		if err := grpcSyncer.sendLog(&entry, nil); err != nil {
-			logger.Error("Error when sending logs to server", zap.Error(err))
+		if shouldBuffer {
 			grpcSyncer.bufferLogEntry(&entry)
```

Comment:
Doing this after unlocking doesn't work, because `bufferLogEntry` doesn't lock, so this new logic is not thread-safe.
The lock must be held when calling `bufferLogEntry`.
What is the reason for moving outside of the Lock/Unlock section?